discovery/discovery.go

```diff
@@ -129,30 +126,52 @@ func (o *orchestratorPool) GetOrchestrators(ctx context.Context, numOrchestrator
 		uris[i] = linfos[j].URL
 	}
 
-	for _, uri := range uris {
-		go getOrchInfo(uri)
-	}
+	var infos []*net.OrchestratorInfo
+	var suspendedInfos *suspensionQueue
+	var timedOut bool
+	var nbResp int
+
+	// try to discover orchestrators until at least 1 is found (with the exponential backoff timout)
+	timeout := getOrchestratorsCutoffTimeout
+	for {
+		ctx, cancel := context.WithTimeout(clog.Clone(context.Background(), ctx), timeout)
+		infoCh := make(chan *net.OrchestratorInfo, numAvailableOrchs)
+		errCh := make(chan error, numAvailableOrchs)
+
+		for _, uri := range uris {
+			go getOrchInfo(ctx, uri, infoCh, errCh)
+		}
 
-	timeout := false
-	infos := []*net.OrchestratorInfo{}
-	suspendedInfos := newSuspensionQueue()
-	nbResp := 0
-	for i := 0; i < numAvailableOrchs && len(infos) < numOrchestrators && !timeout; i++ {
-		select {
-		case info := <-infoCh:
-			if penalty := suspender.Suspended(info.Transcoder); penalty == 0 {
-				infos = append(infos, info)
-			} else {
-				heap.Push(suspendedInfos, &suspension{info, penalty})
+		timedOut = false
+		suspendedInfos = newSuspensionQueue()
```

Comment:
Just want to confirm that the following is the desired behavior:

- If an O is suspended and responds to the discovery request it will be pushed to the suspension queue and will not be counted as a valid response
- If we timeout, then we reset the suspension queue during the next discovery round
- If we hit the max timeout, *then* we will consider using the suspended O in the code block starting on L176
- We end up sending discovery requests to suspended Os more than once if we do multiple rounds of discovery
- The impact is that B priorities using Os that are slower in discovery over Os that are suspended

This seems reasonable, but just want to double check that it is the expected behavior.

Reply:
Yes, I think it's correct. And believe it's the right thing to do, to prioritize slower Os over suspended Os.